### PR TITLE
adjustments

### PR DESCRIPTION
**Pull request by PavelPolyakov**
1. support reply.redirect() in the next() method
2. support setting the policies apply points manually, by default they stay ```onPreHandler``` and ```onPostHandler```

### CODE REVIEW

**Comment by mark-bradshaw** [comment]
Thanks @PavelPolyakov.  On item #2 You took an interesting approach to being able to switch out the extension points.  It's spurred some thoughts for me.  One thing I would like to preserve is the option to bind policies to any and all extension points, rather than limited to just two, a "pre" and a "post".  I think we need to allow policies to bind to any of the events, so I'll probably pull this PR in but alter it so that policies still need to specify which events to bind to, but extend MrHorse to work with all current Hapi events in the request life cycle.

So, for instance, each policy can indicate that it should bind to onRequest, onPreAuth, onPostAuth, onPreRender, onPostRender, etc.  That would allow for maximum flexibility.

**Comment by PavelPolyakov** [comment]
@mark-bradshaw 
I think, in general, there is no case when some policy should be applied twice.
I can't even came with situation, when you should use *pre* and *post* at the same time. Could you give an example?

But, to make it more general, I think we can do it in the way when it should be stated in the policy - where to apply it. And default application point should be stated for the whole plugin.

What do you think?

**Comment by mark-bradshaw** [comment]
I had some scenarios int he past, but I can't come up with an example right
off the top of my head.  But what I was referring to was that I'd like to
allow MrHorse to be able to have policies that applied to more than just
two events at a time.  Currently it can only do policies on prehandler and
posthandler.  I'd like to make it so that all events can be used for your
policies.  So you might have some policies on preHandler, and some on
preAuth, and others on postHandler, etc.  We should set it up to do all
events, and then each policy can just say where it should run.

So to recap, I'm not so concerned with the option to run a single policy on
multiple events, as just making it super easy to use any of the events that
Hapi provides without needing any additional configuration.

I still think that onPreHandler makes the most sense as a default, but I
have no problem with allowing that to be configurable.

**Comment by mark-bradshaw** [comment]
Also, thanks for the PR.  I'd like to see those modifications we've talked about before merging, but I've added you as a collaborator to the repo.

**Comment by PavelPolyakov** [comment]
@mark-bradshaw 
ok, I understood you.

I would update it in the way so it's possible to set the application point right in the policy, the default one would be ```onPreHandler```.
However, this update, from my point of view, would in its turn led to the fact that we would not keep current ```pre``` and ```post``` policy configuration.
And we would have only something like ```applyPoint``` configuration option.

**Comment by mark-bradshaw** [comment]
So each policy would just have one event it would bind to, and that is set
in the policy itself.  That's fine with me, but I wonder if we can't have
both options just about as easily.  We still need to expand MrHorse to know
about the other life cycle events, but the current code is setup to allow
policies to run on multiple events.  What if each policy could supply
either a single event, or an array of events to run on.  I'm still thinking
of going for maximum flexibility as long as it doesn't become hard to
understand or code.

**Comment by PavelPolyakov** [comment]
@mark-bradshaw 
it's not hard to code, but I can't find any case when you actually need to check the request for one policy during the different events from the lifecycle.

**Comment by mark-bradshaw** [comment]
OK.  Works for me.  Please make the necessary modifications and then merge
the PR when it looks good.

**Comment by PavelPolyakov** [comment]
@mark-bradshaw 
Please, review the current variant, if it's ok - I would update the documentation then.

The concept is like we've discussed.
1. The default application policy is ```onPreHandler```
2. One can set the different default application point for the whole plugin, passing the ```applyPoint``` option during the initialisation.
3. One can set the different application point for the particular policy, setting the configuration parameter ```applyPoint``` when being in the policy: https://github.com/PavelPolyakov/mrhorse/blob/master/test/policies/secondPasses.js

The list of the possible application points is hardcoded.
https://github.com/PavelPolyakov/mrhorse/blob/master/lib/index.js#L10

Looking forward for the feedback.

**Comment by mark-bradshaw** [comment]
Thanks a ton.  I'll review that and get back to you.

**Comment by mark-bradshaw** [comment]
Looks good to me.  Very well done.  We have slightly different code styles, but it's close enough that we probably don't need a style guide for contributions at this moment.  In general I try to follow the [hapi style guide](http://hapijs.com/styleguide), but I don't follow it exactly.  I should probably making a CONTRIBUTING.md file and mention that.

So, here's some follow on items I can think of:

* Add an error warning if you mistype the name of an event
* We need to update the README to show the new functionality
* We need to add a section for upgrading, since this PR represents a breaking change
* You should add a CONTRIBUTORS.md file with your name and contact info, web, whatever so you can get credit for your help, and welcome!
* Are there additional tests that we can add to verify the new functionality, such as making sure that you can add to other apply points, or get an error if you mistype an event name?
* Should we have some special handling on policy adding to ease the transition from the old style of event binding to the new?  if (policy.pre) { // add to onPreHandler }

File: lib/index.js [hunk 1]
```diff
@@ -80,25 +90,17 @@ var loadPolicies = function(server, policyDirectory, next) {
...
-                // connect the posthandler if this is the first post policy
-                if (!data.setPostHandler) {
-                    server.ext('onPostHandler', postHandler);
-                    data.setPostHandler = true;
+                // connect the handler if this is the first pre policy
+                if (!data.setHandlers[applyPoint]) {
+                    server.ext(applyPoint, handlers[applyPoint]);
+                    data.setHandlers[applyPoint] = true;
                 }
             }
```
Comment (mark-bradshaw) [inline lib/index.js]: Can we add an `else` here to throw an error if you ask for an event apply point that doesn't exist in our array?

**Comment by mark-bradshaw** [comment]
Oh, and there are some HoundCI messages scattered throughout indicating some possible code smell issues.  Can you just take a look at them and see if any are worth addressing?

**Comment by PavelPolyakov** [comment]
@mark-bradshaw 
>> Should we have some special handling on policy adding to ease the transition from the old style of event binding to the new? if (policy.pre) { // add to onPreHandler }

I don't think that fallback logic should be there, I would prefer developers would update their policies. And we would not support additional code.

Thanks for the feedback, I would proceed.

**Comment by mark-bradshaw** [comment]
@PavelPolyakov It looks like you're all done.  Looks fine to me.  Feel free to merge unless you have more you want to do.

**Comment by PavelPolyakov** [comment]
@mark-bradshaw 
Hi Mark,

I've merged the pull request and created the release.
However, I don't know how to update the package on the npm side.

Is that something you should do? Or I can do it myself somehow.

**Comment by mark-bradshaw** [comment]
I'll take care of it.  I'm adjusting the README a bit and updating tests.
Since this is a breaking change I'm bumping the version to 1.0.0.

**Comment by PavelPolyakov** [comment]
Thanks,

I'm waiting to include it in the project we're working on.

**Comment by mark-bradshaw** [comment]
OK.  I'm waiting for Travis to verify all the tests, and then I'll publish.

**Comment by mark-bradshaw** [comment]
It's done.  I've pushed v1.0.0 to npm.  Just as a note, I altered the
options.applyPoint to options.defaultApplyPoint so that it was more clear
what that did.

**Comment by PavelPolyakov** [comment]
Should we wait for some time?

Because here: 
https://www.npmjs.com/package/mrhorse

I see 0.0.3 still.

**Comment by mark-bradshaw** [comment]
Yes.  It'll take a bit of time to show up.

**Comment by PavelPolyakov** [comment]
@mark-bradshaw 
Are we sure we pushed everything to the npm in the correct way?
Because the number there is still 0.0.3 :) 

https://www.npmjs.com/package/mrhorse

**Comment by mark-bradshaw** [comment]
It's odd.  I definitely published yesterday.  I tried again this morning to
publish, but it said I couldn't overwrite a version already in npm, which
indicates that it does have the new version.

Just for kicks I tried again with a minor version bump, and published again
(npm version minor && npm publish).  The site still isn't updating.  I'm
not sure what's missing or not.

I've contacted npm support.  Hopefully they can clue me in.  If you have
any thoughts please let me know.

Mark

**Comment by mark-bradshaw** [comment]
My second try just went through.  So apparently it's safer just to use the
npm version command to bump versions.  I'll update the build/release
instructions.
